Major refactor

PaulRBerg opened the pull request.

PaulRBerg commented:
The PR is finally ready for review @razgraf @gavriliumircea!!

gavriliumircea reviewed (commented):
The structure looks good. The envio deployment is still not working. I created a draft PR here: https://github.com/sablier-labs/indexers/pull/154 that adds some test that compare the pre and post refactor version of the indexers. I left the comments with the findings. after this are fixed and the latest versions deployed we should rerun the tests.

PaulRBerg commented on src/graph/lockup/mappings/v1.1/SablierV2LockupDynamic.ts:
```diff
@@ -0,0 +1,91 @@
+      shape: null,
+      startTime: params.range.start,
+      streamId: params.streamId,
+      transferable: true,
```
Good catch!

gavriliumircea commented on src/graph/lockup/mappings/common/transfer.ts:
```diff
@@ -0,0 +1,45 @@
+  const index = currentParties.indexOf(currentRecipient);
+  if (index !== -1) {
+    currentParties[index] = newRecipient;
+    stream.parties = currentParties;
```
My bad, there is an issue on sablier-lockup-experimental where the recipient does not get updated when the transfer event occurs. For instance you can check the stream with the id: "0x3a1bea13a8c24c0ea2b8fae91e4b2762a59d7af5-11155111-492". Still investigating will get back once I find the root cause.

PaulRBerg replied:
OK, found the bug — the `parties` is an empty array even for streams that do not have any NFT transfers.

gavriliumircea commented on src/graph/lockup/store/entity-stream.ts:
```diff
@@ -0,0 +1,256 @@
+  }
+
+  /* --------------------------------- STREAM --------------------------------- */
+  stream.alias = Id.streamAlias(chainId, tokenId);
```
Ok fair enough!

PaulRBerg commented:
The Graph indexers are in good shape now.

The only left thing to do is to check the Envio indexers, but we are unable to deploy them at this time — because Envio is not compatible with our repo structure.

Once we do that, @gavriliumircea can run the differential tests (using the tests in PR #154).

gavriliumircea commented:
All the entities on Envio Lockup and Flow have the subgraphId equal to 0 and on Envio Airdrops all the subgraphIds are equal to 1.

PaulRBerg commented:
Fixed the `subgraphId` bug.

PaulRBerg commented on src/envio/airdrops/mappings/common/factory/create-merkle.ts:
```diff
@@ -0,0 +1,93 @@
+  const { context, event, loaderReturn, params } = input;
+
+  /* -------------------------------- CAMPAIGN -------------------------------- */
+  const createEntities = await createAssociatedEntities(context, event, loaderReturn, params);
```
We're not missing the functionality, but it might not be working correctly:

https://github.com/sablier-labs/indexers/blob/8ec320ac333cae42067def7df53a336871cf1b9a/src/envio/airdrops/mappings/v1.2/SablierV2MerkleLockupFactory/create-ll.ts#L6-L13

Will debug later.

PaulRBerg replied:
I can't figure out what's wrong. Could you share the ID of some indexed campaigns that we shouldn't index?

Side note — could the bug instead be in the original indexers, i.e., we had not been indexing some campaigns that we should have indexed?

gavriliumircea replied:
Sure here are 4 campaigns that I found that are indexed only on the new version of envio:

1. 0x70d9bda60024ce56349e89d00eb0869954f837bc-11155111
2. 0x2f14b446a04eb62ce3497a555ab0070e9416ba3b-11155111
3. 0x2e23630266067c5be2fe16feeba8c9841ca009ab-11155111
4. 0xfd101a91b014e46e9eff926fb9da4b5350d4c79e-11155111

They are the first 4 campaigns on sepolia ordered based on the subgraph Id asc

PaulRBerg commented on src/envio/common/store/entity-watcher.ts:
```diff
@@ -0,0 +1,27 @@
+    chainId: BigInt(chainId),
+    id: chainId.toString(),
+    logs: [],
+    streamCounter: 0n,
```
Alright

gavriliumircea commented on src/envio/lockup/store/entity-stream.ts:
```diff
@@ -0,0 +1,265 @@
+    renounceTime: undefined,
+    sender,
+    shape: params.shape ? sanitizeString(params.shape) : undefined,
+    startTime: now,
```
we are in the lockup store not envio store right?

PaulRBerg replied:
whoops, I was mistakenly looking at the Flow mappings locally.

razgraf commented on src/envio/common/id.ts:
```diff
@@ -0,0 +1,75 @@
+   * 0xf50760d8ead9ff322631a1f3ebf26cc7891b3708-137
+   */
+  export function campaign(chainId: number, campaignAddress: Envio.Address): string {
+    return `${chainId}-${campaignAddress.toLowerCase()}`;
```
It is a breaking change in the UI. 

It will require work in the app for SEO purposes (canonical links), since campaign links have already been indexed using the `address-chain` identifiers.

PaulRBerg replied:
hmm, technically, we could refactor the app to resolve the ID first, but I agree that it'd be a better UX to have the app's URL match the indexer's ID.

I will refactor The Graph's campaign ID to start with the campaign address.

PaulRBerg commented on src/envio/lockup/mappings/common/cancel-stream.ts:
```diff
@@ -0,0 +1,40 @@
+    amountB: event.params.recipientAmount,
+    category: "Cancel",
+    streamId: stream.id,
+  });
```
Resolved now, will push a deployment shortly

PaulRBerg commented on src/envio/lockup/mappings/common/create-stream.ts:
```diff
@@ -0,0 +1,104 @@
+  const { entities, rpcData } = loaderReturn;
+  return {
+    asset:
+      entities.asset ?? (await CommonStore.Asset.create(context, event.chainId, params.asset, rpcData.assetMetadata)),
```
Yeah, it looks like the `name` and the `symbol` get set to `[Object object]`. Will fix shortly.

```json
{
  "data": {
    "Stream_by_pk": {
      "id": "0x3a1bea13a8c24c0ea2b8fae91e4b2762a59d7af5-11155111-534",
      "asset": {
        "id": "11155111-0x332DBB4D716D63d88aF8FE51D4b84729489778Ab",
        "name": "[object Object]",
        "symbol": "[object Object]"
      },
      "assetDecimals": "18"
    }
  }
 ```

PaulRBerg replied:
The results had to be accessed as `results[1].result` and `results[2].result`:

```typescript
const metadata = { decimals, name: String(results[1].result), symbol: String(results[2].result) };
```

PaulRBerg commented:
Most (if not all) bugs should be fixed now. New deployments:

- Airdrops: https://indexer.hyperindex.xyz/ca25e55/v1/graphql
- Flow: https://indexer.hyperindex.xyz/4ad8eaf/v1/graphql
- Lockup: https://indexer.hyperindex.xyz/ca25e55/v1/graphql

gavriliumircea commented on src/envio/lockup/mappings/common/cancel-stream.ts:
```diff
@@ -0,0 +1,45 @@
+    streamId: stream.id,
+  });
+  updatedStream = {
+    ...stream,
```
here it should be ...updatedStream instead of ...stream, because of this you are losing the initial changes.

gavriliumircea replied:
Fixed in https://github.com/sablier-labs/indexers/commit/2311d15554981dfcc95a118e2d507cb3caa5db7b

gavriliumircea commented on src/envio/flow/mappings/common/void-stream.ts:
```diff
@@ -0,0 +1,62 @@
+    streamId: stream.id,
+  });
+  updatedStream = {
+    ...stream,
```
here it should be ...updatedStream instead of ...stream

gavriliumircea commented on src/envio/flow/mappings/common/pause-stream.ts:
```diff
@@ -0,0 +1,52 @@
+    streamId: stream.id,
+  });
+  updatedStream = {
+    ...stream,
```
Fixed in https://github.com/sablier-labs/indexers/commit/2311d15554981dfcc95a118e2d507cb3caa5db7b

gavriliumircea commented on src/envio/flow/mappings/common/restart-stream.ts:
```diff
@@ -0,0 +1,54 @@
+    streamId: stream.id,
+  });
+  updatedStream = {
+    ...stream,
```
Fixed in https://github.com/sablier-labs/indexers/commit/2311d15554981dfcc95a118e2d507cb3caa5db7b

gavriliumircea commented on src/envio/airdrops/mappings/common/campaign/clawback.ts:
```diff
@@ -0,0 +1,62 @@
+import type {
+  SablierV2MerkleStreamerLL_v1_1_Clawback_handler as Handler_v1_1,
+  SablierV2MerkleLL_v1_2_Clawback_handler as Handler_v1_2,
+  SablierMerkleInstant_v1_3_Clawback_handler as Handler_v1_3,
```
Fixed in https://github.com/sablier-labs/indexers/pull/92/commits/426a4e0ed35a3f31e4ebf5008a1650488aa6cd8f

gavriliumircea commented on src/envio/airdrops/store/entity-campaign.ts:
```diff
@@ -0,0 +1,172 @@
+): Promise<Entity.Campaign> {
+  let campaign = await createBaseCampaign(context, event, entities, params);
+  const lockupCampaign = createLockupCampaign(params);
+  campaign = {
```
Fixed in [ea68f03](https://github.com/sablier-labs/indexers/pull/92/commits/ea68f037e5b4ce02c657478f1f785cfe7dc43921)

gavriliumircea replied:
Fixed in [ea68f03](https://github.com/sablier-labs/indexers/pull/92/commits/ea68f037e5b4ce02c657478f1f785cfe7dc43921)

PaulRBerg commented on src/envio/airdrops/mappings/v1.3/SablierMerkleFactory/create-lt.ts:
```diff
@@ -49,6 +49,10 @@ struct ConstructorParams {
 
 Contract.Factory.MerkleFactory_v1_3.CreateMerkleLT.handlerWithLoader({
   handler: async ({ context, event, loaderReturn }) => {
+    const lockupAddress = event.params.lockup;
+    if (!isOfficialLockup(event, lockupAddress)) {
+      return;
+    }
```
This should be put in the `common/factory/*` so that the logic is reused

gavriliumircea commented on src/envio/airdrops/mappings/common/campaign/claim-lockup.ts:
```diff
@@ -0,0 +1,91 @@
+    category: "Claim",
+    claimAmount: event.params.amount,
+    claimIndex: event.params.index,
+    claimRecipient: event.params.recipient,
```
This should be lowercased to mirror the graph behaviour

PaulRBerg replied:
Yes but to keep the code DRY, the lowercasing should only be done in the `src/envio/airdrops/store/entity-action.ts` file. Like so:

```
claimRecipient: params.claimRecipient?.toLowerCase(),
```

gavriliumircea commented on src/envio/airdrops/store/entity-campaign.ts:
```diff
@@ -0,0 +1,174 @@
+  // Some fields are set to 0/ undefined because they are set later depending on the campaign category.
+  const campaign: Entity.Campaign = {
+    address: params.campaignAddress,
+    admin: params.admin,
```
This should be lowercased to mirror the graph behaviour

gavriliumircea commented on src/envio/airdrops/store/entity-campaign.ts:
```diff
@@ -0,0 +1,174 @@
+  /* -------------------------------- CAMPAIGN -------------------------------- */
+  // Some fields are set to 0/ undefined because they are set later depending on the campaign category.
+  const campaign: Entity.Campaign = {
+    address: params.campaignAddress,
```
This should be lowercased to mirror the graph behaviour

gavriliumircea commented on src/envio/common/id.ts:
```diff
@@ -0,0 +1,90 @@
+   * 137-0x2791bca1f2de4661ed88a30c99a7a9449aa84174
+   */
+  export function asset(chainId: number, assetAddress: string): string {
+    return `${chainId}-${assetAddress}`;
```
This should be lowercased to mirror the graph behaviour

gavriliumircea commented on src/envio/airdrops/store/entity-campaign.ts:
```diff
@@ -0,0 +1,174 @@
+
+function createLockupCampaign(params: Params.CreateCampaignLockup): Partial<Entity.Campaign> {
+  return {
+    lockup: params.lockup,
```
This should be lowercased to mirror the graph behaviour

gavriliumircea commented on src/envio/common/store/entity-action.ts:
```diff
@@ -0,0 +1,35 @@
+    block: BigInt(event.block.number),
+    category: params.category,
+    chainId: BigInt(event.chainId),
+    contract: event.srcAddress,
```
This should be lowercased to mirror the graph behaviour

gavriliumircea commented on src/envio/lockup/store/entity-stream.ts:
```diff
@@ -0,0 +1,266 @@
+    cliff: false,
+    cliffAmount: 0n,
+    cliffTime: 0n,
+    contract: event.srcAddress,
```
This should be lowercased to mirror the graph behaviour

gavriliumircea commented on src/envio/flow/store/entity-stream.ts:
```diff
@@ -0,0 +1,67 @@
+    batch_id: batch.id,
+    category: "Flow",
+    chainId: BigInt(event.chainId),
+    contract: event.srcAddress,
```
This should be lowercased to mirror the graph behaviour

gavriliumircea commented on src/envio/lockup/store/entity-stream.ts:
```diff
@@ -0,0 +1,266 @@
+  const baseStream = await createBase(context, event, entities, params);
+
+  let initial = {
+    initial: false,
```
the default values are different from the graph, where they are null

gavriliumircea commented on src/envio/lockup/store/entity-stream.ts:
```diff
@@ -0,0 +1,266 @@
+    funder,
+    hash: event.transaction.hash,
+    id: streamId,
+    initial: false,
```
the default values are different from the graph, where they are null

gavriliumircea commented on src/envio/lockup/store/entity-stream.ts:
```diff
@@ -0,0 +1,266 @@
+    depositAmount: params.depositAmount,
+    duration: params.endTime - params.startTime,
+    endTime: params.endTime,
+    funder,
```
This should be lowercased to mirror the graph behaviour

gavriliumircea commented on src/graph/lockup/store/entity-batcher.ts:
```diff
@@ -0,0 +1,16 @@
+import * as Entity from "../bindings/schema";
+
+export function getOrCreateBatcher(sender: Address): Entity.Batcher {
+  const id = sender.toHexString();
```
in envio the batcher id has chainId as well. We should decide what option we want to have

PaulRBerg replied:
With the chain ID

gavriliumircea commented on src/graph/common/id.ts:
```diff
@@ -0,0 +1,85 @@
+  export function factory(factoryAddress: Address): string {
+    const chainId = readChainId().toString();
+    const address = factoryAddress.toHexString();
+    return `${address}-${chainId}`;
```
For envio the id starts with chainId here it is the other way around

gavriliumircea commented on src/graph/airdrops/store/entity-campaign.ts:
```diff
@@ -0,0 +1,148 @@
+
+export function updateCampaignClawback(event: ethereum.Event, campaign: Entity.Campaign, actionId: string): void {
+  campaign.clawbackAction = actionId;
+  campaign.clawbackTime = event.block.timestamp;
```
the field clawbackTo is not set

PaulRBerg commented:
OK I think this is done now. I've fixed all issues and made the equivalent tests pass. I've also refactored the equivalence tests to be DRY.

The new indexers are working, on both Envio and The Graph.

please review @gavriliumircea @razgraf

PaulRBerg commented on src/graph/globals.d.ts:
```diff
@@ -0,0 +1 @@
+/// <reference path="../../node_modules/assemblyscript/std/assembly/index.d.ts" />
```
good catch — installing `assemblyscript` does fix the problem

Although it may not matter, because AssemblyScript is a joke anyway, and I don't think this file achieves anything

PaulRBerg commented on tests/cron/vendor-equivalence/equivalence.ts:
```diff
@@ -0,0 +1,156 @@
+
+        totalCount += envioEntities.length;
+
+        if (envioEntities.length > 0 && envioEntities.length === envioVariables.first) {
+          const nextId = _.toNumber(envioEntities[envioEntities.length - 1].subgraphId);
+          envioVariables.where.subgraphId = { _gt: nextId };
+          graphVariables.where.subgraphId_gt = nextId;
+          return;
+        }
+
+        done = true;
```
Good catch. We can use `continue;` instead of `return;`

PaulRBerg commented on tests/cron/vendor-equivalence/equivalence.ts:
```diff
@@ -86,12 +86,24 @@ async function fetchEntities(
+
+  // Case: $FUEL updated token symbol and name
+
+  if (chainId === "11155111" && asset.toLowerCase() === "0xbaca88a993d9a1452402dc511efeecda1b18c18f") {
+    _.unset(entity, "asset.name");
+    _.unset(entity, "asset.symbol");
+  }
```
Shouldn't we fix the actual problem in the mappings instead of doing this?

razgraf commented:
Mircea and I have concluded our review and all of the issues we found. We'll finish the interfaces review first and then merge and re-deploy.

Two questions @PaulRBerg:
1. I still have the issue of imports not being handled correctly. From what I can see, we'd need to mark the package as a `type: module` in package.json and add the following configuration `"module": "ESNext", "moduleResolution": "node"` in tsconfig. This would make the entire repository "import friendly". Any reason why we shouldn't?

   <img width="915" height="145" alt="Screenshot 2025-07-30 at 12 27 29" src="https://github.com/user-attachments/assets/61de19e5-dbaf-45de-bd2d-2eec8a1240b0" />

2. When I run `just full-check` the generated files (bindings) are also considered, which causes the lint step to fail. I don't think this is the correct behavior, let me know what you think.

razgraf commented:
Thank you! I'm using Node v22.12.0. Will check again pre-merge, but yes good idea isolating it in a separate PR.

My CLI still fails but that's probably still due to linting errors, so I can't say for sure the first probleme doesn't also affect the scripts. Running `just full-check` still shows lots of errors from bindings even after pulling your last commit.

<img width="723" height="76" alt="Screenshot 2025-07-30 at 17 37 16" src="https://github.com/user-attachments/assets/df978645-ec42-465c-bc7c-696b66bebf9d" />

Does yours complete the run successfully?

PaulRBerg commented:
@razgraf you should delete the `src/graph` and `src/envio` directories. They were moved to the root.
